enhancements/olm-deprecate-copied-csvs.md

```diff
@@ -0,0 +1,207 @@
+
+### Risks and Mitigations
+
+#### Info Leak
+
+**Risk:** Operator status can contain references to resources in namespaces a reader of a Copied CSV wouldn't otherwise be made aware of (be privileged to see); i.e. info leak.
+**Partial Mitigation:** Ensure OLM grants `GET` privileges to specific `Operators` based on their scope as defined by the namespaces the related `OperatorGroup` targets; at minimum, this means namespace admins and editors in these target namespaces.
+
+#### OpenShift Console Integration
+
+**Risk:** The OpenShift Console uses Copied CSVs in the rendering of certain pages.
+**Mitigation:** Consult Console maintainers prior to deprecating Copied CSVs.
+
+## Design Details
+
+### Graduation Criteria
+
+#### Dev Preview -> Tech Preview
+
+- Deprecation of Copied CSVs announced; removal within two minor releases
+- Feature toggle introduced, with Copied CSVs enabled by default 
+- End user documentation, relative API stability
+- Sufficient test coverage
+
+#### Tech Preview -> GA
+
+- Copied CSVs disabled by default
+- More testing (upgrade, downgrade, scale)
+- Sufficient time for feedback
+- Feature toggle removed
+
+### Upgrade / Downgrade Strategy
+
+If applicable, how will the component be upgraded and downgraded? Make sure this
+is in the test plan.
+
+Upgrade expectations:
+- Each component should remain available for user requests and
+  workloads during upgrades. Ensure the components leverage best practices in handling [voluntary disruption](https://kubernetes.io/docs/concepts/workloads/pods/disruptions/). Any exception to this should be
+  identified and discussed here.
+- Micro version upgrades - users should be able to skip forward versions within a
+  minor release stream without being required to pass through intermediate
+  versions - i.e. `x.y.N->x.y.N+2` should work without requiring `x.y.N->x.y.N+1`
+  as an intermediate step.
+- Minor version upgrades - you only need to support `x.N->x.N+1` upgrade
+  steps. So, for example, it is acceptable to require a user running 4.3 to
+  upgrade to 4.5 with a `4.3->4.4` step followed by a `4.4->4.5` step.
+- While an upgrade is in progress, new component versions should
+  continue to operate correctly in concert with older component
+  versions (aka "version skew"). For example, if a node is down, and
+  an operator is rolling out a daemonset, the old and new daemonset
+  pods must continue to work correctly even while the cluster remains
+  in this partially upgraded state for some time.
+
+## Implementation History
+
+- N/A
+
+## Drawbacks
+
+- The Operator API is not a "drop-in-place" for Copied CSVs
+- There is no equivalent to namespace scoped CSV `List` requests for `Operator` resources; i.e. users must perform individual `Get` requests for each `Operator` they want to inspect, and must comb through `Operator` status to determine if an operator can act on a target namespace.
```

Comment:
How do users know what operators exist for them to GET? Do they list the subscriptions and get the associated operator for each subscription?

If users can't list, I assume they can't watch. Console uses watch on CSVs in many places in the UI so that pages live update. We'll have to assess the impact and possibly fall back to polling.

Reply:
> How do users know what operators exist for them to GET?

Great question! ...to which I don't have a great answer.

This is essentially the operator discoverability problem, where a user asks: what operators can I use?

The rough edge I'm bumping into answering this is that we're currently trying to drop the concept of "scope" from OLM -- the idea is that operator authors define a set of resources that need to exist on a cluster to activate it for users, groups, namespaces, etc; i.e. `OperatorGroups` are being replaced by per-operator `Templates` -- as a result, the concept of "availability" in a namespace doesn't mean the same thing from operator to operator.

Without the "scopeless" world completely fleshed out, it becomes really hard to come up with a solution here that doesn't just lean on the existing concepts of "scope"; e.g. project the list of "watching" operators onto namespace annotations.

I wonder if we could think about this from another perspective. I'm really curious to know how the pages that rely on Copied CSVs are being used. For example, if 99% of users only need to know how to use an API they are privileged enough to interact with, can we hoist that descriptor metadata out of CSVs (to some shared resource like its CRD)? Can we change the question from "what operators can interact with" to "what APIs do I have access to and how do I use them?".

Reply:
OR is there a way that the pages that do rely on the copied csv can detect with they are disabled and put up some kind of big warning message that indicates that the csv copying has been disabled due to performance issues/etc and that you need to do something else to get the information the user is looking for?  (like go have an admin go do something for you)

Reply:
@njhale I asked a _very_ similar question in the OLM F2F and the answer from @dmesser was that CRDs can be too low-level (e.g. if an operator provides 25 APIs in various groups, how do you know what operator they came from so you can go read the operator docs that tell you want to do?)

Reply:
> I asked a very similar question in the OLM F2F and the answer from @dmesser was that CRDs can be too low-level

@joelanford I remembered this, but I don't think I'm quite ready to wave my hands there yet, which is why I'm asking here.

> e.g. if an operator provides 25 APIs in various groups, how do you know what operator they came from so you can go read the operator docs that tell you want to do?

Assuming we do settle on the Operator for discovery, I'm struggling to think of alternatives to projecting "watched operators" down into a namespace's topology -- even that still builds into our current model of "scope".

Reply:
For now, I've eschewed answering this question by reducing the scope of this EP. Please see the new title and description.